### ISSUE TITLE
PartialLoader: $translate.use() and $translate.refresh() in race-condition, wrong language loaded

### ISSUE THREAD

**Issue by SimonSimCity** [angular-translate/angular-translate#1781]
### Subject of the issue
If I call `$translate.use()` and add another partial template directly after that, this one won't get loaded in the new language. 

### Your environment
* version of angular-translate: 2.15.2
* version of angular: 1.5.11
* which browser and its version: Every ...

### Steps to reproduce

https://www.webpackbin.com/bins/-Kv2DNtdpnsNYAU6gMTA

1. Set up a project using the [`partialLoader`](https://angular-translate.github.io/docs/#/guide/12_asynchronous-loading) having at least two languages (f.e. English and German).
1. Set English as fallback language
1. Call `$translate.refresh()` every time your parts change:
```js
  $rootScope.$on('$translatePartialLoaderStructureChanged', function () {
    $translate.refresh();
  });
```
4. Let your app start off by choosing German as language (by `$translate.use('de')`) and in the next line add a new part.

### Expected behaviour
The new part should be available in German

### Actual behaviour
The new part is only loaded in English, the old (in this case the fallback) language.

### Explanation
When calling `$translate.use('de')`, the language effectively hasn't changed, but it has a callback, that is called when the change of the language was successful. This creates a problem in this case here, because when now calling `$translatePartialLoader.addPart('test')` and thereby `$translate.refresh()`, before the callback of `$translate.use('de')` comes back, the value of the internal variable `$uses` is still on the previous value, and therefore the call of `$translate.refresh()` is called before the value is changed. This method in turn will reload all the available translations after emptying them. This will also overwrite the content of the `$translate.use('de')` callback by the time it comes back.

### A workaround
Call `$translate.refresh()` again after `$translation.use('de')` has finished. Use the promise like shown here:
```js
$translation.use('de').then(() => $translate.refresh());
```

The question of me now:

Does all this `$translate.refresh()` pull down on my performance? I haven't run tests against it, but I could well imagine that this (having quite some parts loaded) can take on performance ...

**Comment by SimonSimCity** [comment]
Could well be a duplicate of #1704, #1703 ... The issue #1703 can at least be solved by the work-around I provided here.

**Comment by knalli** [comment]
@SimonSimCity  We could consider to call such a refresh implicitly (for the user). Did you have see some (performance) issues with this? 

Maybe you want make a PR for this?

**Comment by knalli** [comment]
I have added a watcher while fetching all parts, but it is limited to a magic value of `3 retries`. This would prohibit loops.

* https://plnkr.co/edit/aUqSfm?p=preview by #1703 @detollenaere
* https://www.webpackbin.com/bins/-Kyh4BSpIjpsSJmsWCfu by @SimonSimCity 

What do you think?

**Comment by knalli** [comment]
See also ad9213d

**Comment by SimonSimCity** [comment]
Was about to check it out but only got a 500-error when looking for `https://run.plnkr.co/6n3Uc9MttusynxQV/angular-translate.2.17-dev.js` ...

**Comment by knalli** [comment]
try again

**Comment by SimonSimCity** [comment]
@knalli seems to work ... but this would now have to be implemented in every loader that acts asynchronously, and in addition: You'll still have the potential of a race-condition in the `use()` method if it's used by another service ...

**Comment by knalli** [comment]
Why? The problem exists in partial loader only. What have I missed?

**Comment by SimonSimCity** [comment]
To me it doesn't matter what component the problem was in combination with, but rather the source of it.

The source is (at least as of how I see it), that in the `$translate.use()` method, you only set the language _after_ a promise has returned. ([see here](https://github.com/angular-translate/angular-translate/blob/master/src/service/translate.js#L1931))

In your implementation of `use()`, every language loader is running using a promise. Means, that when using a translation-provider, that uses a promise here, these two results on the console will differ:

```js
$translate.use('es').then((lang) => console.log(lang));
console.log($translate.use());
```

And it won't matter how long the promise will delay the result, because it's anyways just executed after the current thread has finished. For this reason, every interaction, that happens before the promise resolves, will still be using the old language. Hope this clarifies what I am up for. This also yields for the case when there's no async action done by the promise. In this case, the promise will resolve once the process is idle. Means, that if you have  another promise you registered earlier, it will be executed before this.

Try the following in any angular project (or get $q from somewhere else ...):

```js
var b = 1;
$q.all().then(() => { console.log(b); });
$q.all().then(() => { b = 2; });

console.log(b);
```

**Comment by knalli** [comment]
Well, that's right. `$translate.use()` returns the current, active language. Anything else would be incorrect as being something which *could be* in the future. Therefore, the difference is correct. That's also why adding a part *after* a successful language change would not cause this issue even without the proposed patch/workaround. 

It is recommended to modify parts before a language change or after a completed change. But the order cannot be guaranteed by you (as an app developer) always, because the part could be added in a different place without knowing of a language change. That's why we need this additional workaround.

Speaking of the loaders: only the partial loader has an additional internal structure (modified by adding and removing parts at any time) which have a side-effects for active loadings. But the other loaders do not have such side-effects. Maybe there will be 2 loads in parallel, but never for the same language.

**Comment by knalli** [comment]
Technical background: Only the `$translate` service invokes the loaders and could handle the responses of them. That is why the partial loader cannot actively load itself (i.e. loading the parts just in time they have been added, and making addPart currently-loading-aware). We would need an additional back-channel on which loaders could notify they have been outdated.

The workaround patches the current active request and intercepts an additional refresh for missing parts added between load start and finish.

**Comment by knalli** [comment]
fyi: just picked the code above in the current canary branch. Hold for next release.
